packages/editor/src/optimizer/optimizer.ts

```diff
@@ -0,0 +1,254 @@
+import {
+  isPenroseError,
+  LabelMeasurements,
+  PenroseError,
+  runtimeError,
+} from "@penrose/core";
+import consola from "consola";
+import { Result } from "true-myth";
+import {
+  collectAndSeparateLabels,
+  CompileRequestData,
+  CompileResult,
+  ComputeLayoutRequestData,
+  DiagramID,
+  DiscardDiagramData,
+  HistoryLoc,
+  InvalidDiagramIDError,
+  InvalidHistoryLocError,
+  isErr,
+  layoutStateToRenderState,
+  logLevel,
+  MessageID,
+  MessageIDGenerator,
+  MessageRequestData,
+  MessageResponse,
+  MessageResult,
+  MessageTags,
+  Notification,
+  notifyWorker,
+  PollRequestData,
+  PollResult,
+  RenderState,
+  request,
+  ResampleRequestData,
+  ResampleResult,
+  resolveResponse,
+  showOptimizerError,
+  spinAndWaitForInit,
+  Tagged,
+  taggedErr,
+} from "./common.js";
+
+const log = consola.create({ level: logLevel }).withScope("optimizer:client");
+
+export default class Optimizer {
+  private resolvesByMsgId = new Map<
+    MessageID,
+    (result: MessageResult) => void
+  >();
+
+  private svgCaches = new Map<DiagramID, Map<string, HTMLElement>>();
+  private labelMeasurements = new Map<DiagramID, LabelMeasurements>();
+
+  private messageIdGenerator = new MessageIDGenerator();
+  private broker: Worker;
+
+  /**
+   * Create an `Optimizer` with an already initialized broker. Private since
+   *  users should just call `Optimizer.create()`.
+   * @param broker Initialized broker
+   * @private
+   */
+  private constructor(broker: Worker) {
+    this.broker = broker;
+    broker.onmessage = this.onBrokerMessage.bind(this);
+    log.info(`Optimizer created with initialized broker`);
+  }
+
+  private onBrokerMessage = ({
+    data,
+  }: MessageEvent<MessageResponse | Notification>) => {
+    switch (data.tag) {
+      case "MessageResponse":
+        log.info(`Main thread received response ${data.result.tag}`);
+        resolveResponse(data, this.resolvesByMsgId);
+        break;
+
+      case "Notification":
+        throw new Error(
+          `Optimizer received unknown notification ${data.data.tag}`,
+        );
+    }
+  };
+
+  /** Convenience wrapper */
+  private request = <T, R extends MessageResult & Tagged<T>>(
+    requestData: MessageRequestData & Tagged<T>,
+  ): Promise<R> => {
+    return request<T, R>(
+      this.broker,
+      requestData,
+      this.resolvesByMsgId,
+      this.messageIdGenerator,
+    );
+  };
+
+  /**
+   * Create an optimizer.
+   */
+  static create = async (): Promise<Optimizer> => {
+    log.info("Spinning new broker");
+    return new Optimizer(await spinAndWaitForInit("./broker.ts"));
```

Comment:
I'm slightly worried about whether this relative URL will always get correctly transformed, since it's not being used in a literal `new URL` expression. Could you change this so it instead uses an [`import` with Vite's `?worker` suffix](https://v3.vitejs.dev/guide/features.html#import-with-query-suffixes)?

Reply:
It is being transformed, within `spinAndWaitForInit`.